src/expr/explain.rs

```diff
@@ -0,0 +1,417 @@
+// Copyright Materialize, Inc. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+/// This is the implementation for the EXPLAIN command.
+///
+/// Conventions:
+/// * nodes in the tree are referred to by their index in the post-order traversal, written eg "%4"
+/// * columns are written eg "#4"
+/// * collections of columns are written as ranges where possible eg "#2..#5"
+use super::{AggregateExpr, Id, IdHumanizer, LocalId, RelationExpr, ScalarExpr};
+use std::collections::HashMap;
+
+#[derive(Debug)]
+pub struct Explanation<'a> {
+    /// One ExplanationNode for each RelationExpr in the plan, in left-to-right post-order
+    pub nodes: Vec<ExplanationNode<'a>>,
+}
+
+#[derive(Debug)]
+pub struct ExplanationNode<'a> {
+    /// The expr being explained
+    pub expr: &'a RelationExpr,
+    /// The parent of expr
+    pub parent_expr: Option<&'a RelationExpr>,
+    /// A pretty-printed representation of the expr
+    pub pretty: String,
+    /// A list of annotations containing extra information about this expr
+    pub annotations: Vec<String>,
+    /// Nodes are grouped into chains of linear operations for easy printing
+    pub chain: usize,
+}
+
+impl<'a> std::fmt::Display for Explanation<'a> {
+    fn fmt(&self, f: &mut std::fmt::Formatter) -> Result<(), std::fmt::Error> {
+        write!(
+            f,
+            "\
+RelationExprs are printed in post-order, left to right
+RelationExprs which only have a single input are grouped together
+Each group of RelationExprs is referred by id eg %4
+Columns are referred to by position eg #4
+RelationExprs may be followed by additional annotations on lines starting with | |
```

Comment:
I would much prefer that we not include these instructions in every result set. No other database that I know of does this, and it will interfere with folks who want to programmatically inspect the output of EXPLAIN.

If we do think it's essential to humans, we could send along a pgwire NOTICE message, which is a more appropriate place to put stuff like this.

Reply:
:+1: